File: app/src/main/java/edu/anbu/libraryofalexandria/SearchBookActivity.java
package edu.anbu.libraryofalexandria;

import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.appcompat.widget.SearchView;
import android.view.MenuItem;
import android.view.View;

import java.util.ArrayList;
import java.util.List;

public class SearchBookActivity extends AppCompatActivity implements BookItemClickListener {

    private BookAdapter mSearchBookAdapter;
    private List<Book> searchResultList;
    private SearchAsyncTask searchAsyncTask;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_search_book);

        androidx.appcompat.widget.Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        ActionBar supportActionBar = getSupportActionBar();
        if (supportActionBar != null) {
            supportActionBar.setDisplayHomeAsUpEnabled(true);
        }

        SearchView searchView = findViewById(R.id.search_book_search_view);
        RecyclerView searchResultRecyclerView = findViewById(R.id.search_result_recycler_view);

        searchResultList = new ArrayList<>();
//        addDummyList();
        searchResultRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        mSearchBookAdapter = new BookAdapter(this, searchResultList, true);
        searchResultRecyclerView.setAdapter(mSearchBookAdapter);

        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                onSearch(query);
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                onSearch(newText);
                return false;
            }
        });
    }

    @Override
    public void onItemClicked(View v, int position) {
        Intent intent = new Intent(this, BookDetailActivity.class);
        intent.putExtra("Book", searchResultList.get(position));
        startActivity(intent);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
        }
        return super.onOptionsItemSelected(item);
    }

    private void onSearch(String keyword) {
        if (keyword == null || keyword.trim().equals("")) {
            searchResultList.clear();
//            mSearchBookAdapter.updateItems();
            return;
        }
        if (searchAsyncTask != null) {
            searchAsyncTask.cancel(true);
        }
        searchAsyncTask = new SearchAsyncTask(bookList -> {
            searchResultList.clear();
            searchResultList.addAll(bookList);
//            mSearchBookAdapter.updateItems();
        });
        searchAsyncTask.execute(keyword.trim());
    }

//    private void addDummyList() {
//        Book book1 = new Book("The Java Programming Language", 1, "James Gosling and ken Arnold", 2, "Programming foundation", "genre", "This book an introduce to programming in Oracle's Java programming language, a widely used programming language software platform.", 2017);
//        Book book2 = new Book("The Java Programming Language", 1, "James Gosling and ken Arnold", 2, "Programming foundation", "genre", "This book an introduce to programming in Oracle's Java programming language, a widely used programming language software platform.", 2017);
//        searchResultList.add(book1);
//        searchResultList.add(book2);
//    }

    private static class SearchAsyncTask extends AsyncTask<String, Void, List<Book>> {

        SearchBookListener listener;

        SearchAsyncTask(SearchBookListener listener) {
            this.listener = listener;
        }

        @Override
        protected List<Book> doInBackground(String... strings) {
            return AppDatabase.getInstance().bookDao().search("%" + strings[0] + "%");
        }

        @Override
        protected void onPostExecute(List<Book> books) {
            super.onPostExecute(books);
            listener.onSearchComplete(books);
        }
    }
}

File: app/src/main/java/edu/anbu/libraryofalexandria/SearchBookActivity.kt
package edu.anbu.libraryofalexandria

import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.LinearLayout
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import androidx.appcompat.widget.Toolbar
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.updatePadding
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import edu.anbu.libraryofalexandria.AppDatabase.Companion.instance
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch

class SearchBookActivity : AppCompatActivity(), BookItemClickListener {
    private lateinit var mSearchBookAdapter: BookAdapter
    private var searchResultList: MutableList<Book>? = null
    private var searchJob: Job? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_search_book)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        val supportActionBar = supportActionBar
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        val rootView = findViewById<LinearLayout>(R.id.root_linear_layout)
        ViewCompat.setOnApplyWindowInsetsListener(rootView) { view, listener ->
            val systemBarsInsets = listener.getInsets(WindowInsetsCompat.Type.systemBars())
            view.updatePadding(top = systemBarsInsets.top)
            listener
        }

        val searchView = findViewById<SearchView>(R.id.search_book_search_view)
        val searchResultRecyclerView = findViewById<RecyclerView>(R.id.search_result_recycler_view)
        searchResultRecyclerView.layoutManager = LinearLayoutManager(this)
        mSearchBookAdapter = BookAdapter(this, searchResultList, true)
        searchResultRecyclerView.adapter = mSearchBookAdapter
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                onSearchNew(query)
                return false
            }

            override fun onQueryTextChange(newText: String): Boolean {
                onSearchNew(newText)
                return false
            }
        })
    }

    override fun onItemClicked(v: View?, position: Int) {
        val intent = Intent(this, BookDetailActivity::class.java)
        intent.putExtra("Book", searchResultList?.get(position))
        startActivity(intent)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
        }
        return super.onOptionsItemSelected(item)
    }

    private fun onSearchNew(keyword: String?) {
        searchJob?.cancel()
        if (keyword == null || keyword.trim { it <= ' ' } == "") {
            searchResultList?.clear()
            mSearchBookAdapter.updateItems(null)
            return
        }

        searchJob = lifecycleScope.launch(Dispatchers.IO) {
            val searchKeyword = "%" + keyword.trim { it <= ' ' }.plus("%")
            val books = instance.bookDao().search(searchKeyword)
            searchResultList = books?.toMutableList()
            lifecycleScope.launch(Dispatchers.Main) {
                mSearchBookAdapter.updateItems(searchResultList)
            }
        }
    }
}